subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/FileChange.java

```diff
@@ -18,22 +18,55 @@
 
 import com.google.common.base.Objects;
 import org.gradle.api.tasks.incremental.InputFileDetails;
+import org.gradle.internal.nativeintegration.filesystem.FileType;
 
 import java.io.File;
 
 public class FileChange implements TaskStateChange, InputFileDetails {
     private final String path;
     private final ChangeType change;
     private final String fileType;
+    private final FileType previousType;
+    private final FileType currentType;
 
-    public FileChange(String path, ChangeType change, String fileType) {
+    public static FileChange added(String path, String title) {
+        return new FileChange(path, ChangeType.ADDED, title, null, null);
+    }
+
+    public static FileChange removed(String path, String title) {
+        return new FileChange(path, ChangeType.REMOVED, title, null, null);
```

Comment:
- [ ] Can we somehow not use `null` here? We know that a `REMOVED` "thing" must be `MISSING` by now, and an `ADDED` thing must have been `MISSING` before. We might as well record the file type we have now: we might as well use it to generate the message, so we could say "directory was removed." Even if we don't do that, I think there should be a way to avoid using `null`.

Another solution could be to calculate the displayed change type on the caller side.